### PR TITLE
[TP] Shard activations

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
Let's shard the activations for tensor parallelism as well.

Test Plan:
Tested on v4-8.

### CODE REVIEW

File: examples/pytorch/language-modeling/run_clm.py [hunk 1]
```diff
@@ -507,70 +508,71 @@ def main():
     # host-side OOM
     sharded_state_dict = {}
     for name, param in model.state_dict().items():
-      # Create an tensor based on the meta tensor
```
Comment (alanwaketan) [inline examples/pytorch/language-modeling/run_clm.py:510]: The indentation is weird. So I fix it.

File: src/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -241,7 +241,8 @@ def __init__(self, config: LlamaConfig):
         super().__init__()
         self.config = config
         # For PyTorch/XLA's SPMD 2D sharding
-        self.spmd_2d_sharding = config.spmd_2d_sharding
+        assert config.spmd_2d_sharding & config.spmd_tensor_sharding == 0
```
Comment (jonb377) [inline src/transformers/models/llama/modeling_llama.py:244]: `&` is bitwise `and`, right? So spmd_2d_sharding=2 and spmd_tensor_sharding=4 would pass this test

File: src/transformers/models/llama/modeling_llama.py [hunk 3]
```diff
@@ -582,7 +583,8 @@ class LlamaModel(LlamaPreTrainedModel):
     def __init__(self, config: LlamaConfig):
         super().__init__(config)
         # For PyTorch/XLA's SPMD 2D sharding
-        self.spmd_2d_sharding = config.spmd_2d_sharding
+        assert config.spmd_2d_sharding & config.spmd_tensor_sharding == 0
+        self.spmd_2d_sharding = config.spmd_2d_sharding + config.spmd_tensor_sharding
```
Comment (jonb377) [inline src/transformers/models/llama/modeling_llama.py:587]: small nit: can we rename the field to `self.model_sharding`?

File: src/transformers/models/llama/modeling_llama.py (hunk 2 above)
Comment (alanwaketan) [reply src/transformers/models/llama/modeling_llama.py:244]: Okay, I guess python land doesn't have the C++ &&...

Comment (JackCaoG) [reply src/transformers/models/llama/modeling_llama.py:244]: haha, I made the same mistake before as well

**Comment by alanwaketan** [comment]
Let's close this and move to shard every matmul!

**Pull request closed without merging**